pkg/controller/proxy/controller_test.go

```diff
@@ -0,0 +1,168 @@
+package proxy
+
+import (
+	"context"
+	"testing"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/kubernetes/fake"
+	core "k8s.io/client-go/testing"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+
+	"github.com/ianlewis/controllerutil/informers"
+	"github.com/ianlewis/controllerutil/logging"
+
+	"github.com/ianlewis/memcached-operator/pkg/apis/ianlewis.org/v1alpha1"
+	ianlewisorgfake "github.com/ianlewis/memcached-operator/pkg/client/clientset/versioned/fake"
+	ianlewisorginformers "github.com/ianlewis/memcached-operator/pkg/client/informers/externalversions/ianlewis/v1alpha1"
+)
+
+func getKey(o interface{}, t *testing.T) string {
+	if key, err := KeyFunc(o); err != nil {
+		t.Errorf("Unexpected error getting key for %v: %v", o, err)
+		return ""
+	} else {
+		return key
+	}
+}
+
+func alwaysSynced() bool {
+	return true
+}
+
+type fixture struct {
+	t *testing.T
+
+	client            *fake.Clientset
+	ianlewisorgClient *ianlewisorgfake.Clientset
+
+	proxies []*v1alpha1.MemcachedProxy
+
+	// Actions expected to happen on the client.
+	actions []core.Action
+}
+
+func newFixture(t *testing.T) *fixture {
+	return &fixture{
+		t: t,
+	}
+}
+
+func (f *fixture) newController() (*Controller, informers.SharedInformers) {
+	objects := make([]runtime.Object, len(f.proxies))
+	for i, p := range f.proxies {
+		objects[i] = p
+	}
+
+	f.client = fake.NewSimpleClientset()
+	f.ianlewisorgClient = ianlewisorgfake.NewSimpleClientset(objects...)
+
+	informers := informers.NewSharedInformers()
+	proxyInformer := informers.InformerFor(
+		&v1alpha1.MemcachedProxy{},
+		func() cache.SharedIndexInformer {
+			return ianlewisorginformers.NewMemcachedProxyInformer(
+				f.ianlewisorgClient,
+				corev1.NamespaceAll,
+				0,
+				cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc},
+			)
+		},
+	)
+
+	c := New(
+		"test-proxy-controller",
+		f.client,
+		f.ianlewisorgClient,
+		proxyInformer,
+		&record.FakeRecorder{},
+		logging.New(""),
+		1,
+	)
+
+	c.pListerSynced = alwaysSynced
+	for _, p := range f.proxies {
+		proxyInformer.GetIndexer().Add(p)
```

Comment:
I do this through either `watch.NewFake()` or by adding a Reactor to the fakeClient (depending on the test).

I know Kubernetes code will go to the Indexer directly - do you see any specific benefits? Personally, I've no idea what the Indexer actually does, so it feels nicer to go through the higher level abstractions. But personal preference.

Reply:
Interesting. do you have an example?

AFAIK the indexer is just a cache so fhis just adds it to the cache so that the local listers can get it without a full sync through the client

I mostly just copied this method from the deployment controller's tests tbh though

Reply:
Ah - that's the benefit there! I have to do the full sync before the Informer sees it.

I had seen it in the deployment controller tests as well, but was hesitant to use it, as I wasn't sure what it was doing.

LGTM!